### ISSUE TITLE
give an option for default host and port when building with `adapter-node`

### ISSUE THREAD

**Issue by m4rch3n1ng** [sveltejs/kit#2115]
### Describe the problem

when building an app using `adapter-node` i would like to be able to choose the default port and host

right now if i run `$ svelte-kit build` it automatically chooses `"0.0.0.0"` as the default host and `3000` as the default port and would like to be able to control that (for the build)

the only way that i could find to change the host and port is using

```
$ HOST=localhost PORT=4000 node build
```

to start the app, but that is temporary and only affects the environmental variables and not the defaults and i would have to do that every time i start

if there is a way to specify that and i haven't found it then i'm sorry, if this is a duplicate then i am too.

_iirc there used to be a time where you could specify the `--port [port]` option as you ran `$ svelte-kit build` but i don't know if that actually did anything or if i even remember correctly_

#### config

i would suggest adding the option `default` (or maybe defaults, given that it is `port` *and* `host`), so the config could look something like this

```js
import node from "adapter-node"

export default {
    kit: {
        adapter: node({
            default: {
                host: "localhost",
                port: 4000
            }
        })
    }
}
```

#### implementation

i dont really know a lot about this codebase so i dont actually know how one would implement that but in the `adapter-node` index.js file, lines 58 through 63, where it goes

```js
writeFileSync(
    '.svelte-kit/node/env.js',
    `export const host = process.env[${JSON.stringify(
        host_env
    )}] || '0.0.0.0';\nexport const port = process.env[${JSON.stringify(port_env)}] || 3000;`
);
```

seems like an idea and then after line 37 you could add a line with

```js
    default: { host: default_host = '0.0.0.0', port: default_port = 3000 } = {},
```

~but i don't know if that would work~

after a bit of testing, adding that line and expanding the first-mentioned code to

```js
writeFileSync(
    '.svelte-kit/node/env.js',
    `export const host = process.env[${JSON.stringify(
        host_env
    )}] || ${JSON.stringify(default_host)};\nexport const port = process.env[${JSON.stringify(
        port_env
    )}] || ${JSON.stringify(default_port)};`
);
```

works as expected, a very simple fix to a very simple problem, i would just need some confirmation on whether or not this would be a good addition by the package-maintainers

### Alternatives considered

like i already mentioned the only alternative i found is to start your production server with

```
$ HOST=localhost PORT=4000 node build
```

but i find that tedious and would appreciate if i could just give standard options to the builder

### Importance

nice to have

### Additional Information

_no additional information_

**Comment by cbergen** [comment]
According to the readme here: https://github.com/sveltejs/kit/tree/master/packages/adapter-node this should be possible.

```js
// svelte.config.js
import adapter from '@sveltejs/adapter-node';

export default {
	kit: {
		adapter: adapter({
			// default options are shown
			out: 'build',
			precompress: false,
			env: {
				host: 'localhost',
				port: '4000'
			}
		})
	}
};
```

But given the above settings, my compiled `index.js` file includes this:

```js
// .svelte-kit/node/env.js
var host = process.env["localhost"] || "0.0.0.0";
var port = process.env["4000"] || 3e3;
```

Which doesn't look right. Hacking the compiled index file above to this:
```js
// .svelte-kit/node/env.js
var host = process.env["localhost"] || "0.0.0.0";
var port = process.env["4000"] || 3e3;
host = "localhost";
port = 4000;
```
does work.

**Comment by cbergen** [comment]
Nevermind. I see now that the purpose of the "env" configuration is to set the environment variable name, not the value. I agree though that setting the value would be convenient.

**Comment by m4rch3n1ng** [comment]
> I see now that the purpose of the "env" configuration is to set the environment variable name, not the value

i assumed that would be the case too but then was very confused when it wasn't, so maybe the documentation is lacking a bit of clarity too.

> // .svelte-kit/node/env.js
var host = process.env["localhost"] || "0.0.0.0";
var port = process.env["4000"] || 3e3;
host = "localhost";
port = 4000;

that code would work, but it prohibits using the environmental variables to change the defaults like so:

> $ HOST=localhost PORT=4000 node build

i added a snippet on a proposed change to the code that actually generated that part of the index.js, which would generate a file like

```js
// .svelte-kit/node/env.js
var host = process.env["HOST"] || "localhost";
var port = process.env["PORT"] || 4e3;
```

which would only set the default, instead of always overriding the variable

**Comment by cbergen** [comment]
At least for me, I stumbled on the docs thinking that I could add these variables in my svelte config. I've now moved them to the server as environment variables and that works well enough.

**Comment by jchanes04** [comment]
For me, the process.env is still empty even when I create a .env file in the root directory. I did a quick ctrl + F, and nowhere in the code is process.env assigned to, nor is the dotenv module used to load it. This means that there's no way for the port to be changed from the default of 3000, and I have to go in and manually reassign it. Does anyone else have this issue?

**Comment by m4rch3n1ng** [comment]
> This means that there's no way for the port to be changed from the default of 3000, and I have to go in and manually reassign it. Does anyone else have this issue?

i am sorry for responding so late, but there actually is a way, quoting my original issue:

> like i already mentioned the only alternative i found is to start your production server with
>
> $ HOST=localhost PORT=4000 node build
>
> but i find that tedious and would appreciate if i could just give standard options to the builder

this just sets the needed environment variables for one time

if that doesn't work, the npm package [cross-env](https://npmjs.com/package/cross-env) most definitely should

**Comment by jchanes04** [comment]
Thanks for the fix. This solution works, but isn't very user friendly at all. The adapter should still either read the actual port value from svelte.config.js or load any .env file present at run time. Both have their advantages and disadvantages, but either is an improvement over the current functionality.

**Comment by xpat** [comment]
This is a major issue for me.  I've been trying to figure it out for a while now.  I'm glad I found this thread.  

My goal is to build any number of svelte-kit apps and serve each of them on its own domain on my own vm on Google Cloud Platform using Nginx as a reverse proxy and PM2 to manage the processes. So far no luck. I've only succeeded at serving one app at a time and that same app comes through the same on each of the domains for which I have run `PM2 start build/index.js`  (I outline all the steps I go through to deploy my built app in [a question I answered on StackOverflow](https://stackoverflow.com/a/70571992/14203669).  I included a copy of one of my Nginx configurations in that answer.)

On the pre-build end of things, on my local development machine, I've tried the approaches mentioned by @m4rch3n1ng when they opened the issue, specifically:

1.  add env variables to svelte.config.js
```
import adapter from '@sveltejs/adapter-node';
import preprocess from 'svelte-preprocess';

/** @type {import('@sveltejs/kit').Config} */
const config = {
        // Consult https://github.com/sveltejs/svelte-preprocess
        // for more information about preprocessors
        preprocess: preprocess(),

        kit: {
                adapter: adapter({
                        // default options are shown
                        out: 'build',
                        precompress: false,
                        env: {
                                host: 'localhost',
                                port: '4000'
                        }
                }),

                // hydrate the <div id="svelte"> element in src/app.html
                target: '#svelte'
        }
};
```
2.  run* the build, prepending host and port 
`$ HOST=localhost PORT=4000 npm run build`

* having read through the issue, I noticed that some use **node build** or **svelte-kit build** whereas I've only been able to use **npm run build**, just fyi.  

If I run the build after specifying PORT=4000, there's no change to my build/index.js file; only if I change the env variables in svelte.config.js does my output change:
```
/* global "SOCKET_PATH", "localhost", "4000" */

const path = process.env["SOCKET_PATH"] || false;
const host = process.env["localhost"] || '0.0.0.0';
const port = process.env["4000"] || (!path && '3000');
```

* This is similar to that which @m4rch3n1ng describes. 
```
// .svelte-kit/node/env.js
var host = process.env["localhost"] || "0.0.0.0";
var port = process.env["4000"] || 3e3;
host = "localhost";
port = 4000;
```

In sum, the first step I take is to run npm build after adding a new port value to the svelte.config.js file and then uploading the build and package*.* files to my remote server.  

(Then I run npm install to generate the node_modules folder and move all the files to /var/www/domainname, a different one for each of my apps.)  

I was hoping none of this would be necessary and that all my apps could be built with the default port 3000 and only after uploading it to my vm would I need to specify a new port either on Nginx upstream or using PM2, or at the very least, be able to edit the build/index.js file and adjust the port there, once built and uploaded.  

One minor question I have is whether the port variable ought to be in single quotes if it's numeric -- probably not, right? (either way, it's still not working for me in production.)  It's only if it is a variable name and then you pass the variable name at the command line, as in these examples:  [github adapter-node readme](https://github.com/sveltejs/kit/blob/master/packages/adapter-node/README.md)

**Comment by cbergen** [comment]
This looks like is the same trap I fell into. The svelte.config.js env variables are not intended for setting server environment variable values. They tell svelte which env variable _names_ to use. So in svelte.config.js, 

The default is:
```js
// svelte.config.js
import adapter from '@sveltejs/adapter-node';

export default {
	kit: {
		adapter: adapter({
			// default options are shown
			out: 'build',
			precompress: false,
			env: {
				host: 'HOST',
				port: 'PORT'
			}
		})
	}
};
```

The above tells svelte to look for the env variable HOST and PORT. These must be strings and on your server in your .env or whatever you would use `HOST=localhost PORT=4000`

The only time you would need to change the svelte.config.js env settings is in a case like this:
```js
// svelte.config.js
import adapter from '@sveltejs/adapter-node';

export default {
	kit: {
		adapter: adapter({
			out: 'build',
			precompress: false,
			env: {
				host: 'MYHOST',
				port: 'MYPORT'
			}
		})
	}
};
```
And then starting the node server like:
```
MYHOST=localhost MYPORT=4000 node build
```

Basically the server determines the port, adapter-node can work on whatever port it needs as long as it knows what environment variables to look for.

**Comment by xpat** [comment]
@cbergen Thanks a million for the clarification.  I read your comment and then I ran the build locally after keeping 'HOST' and 'PORT' as names (as you pointed out), changed the port to 4000 in my Ngnix configuration file, and ran $HOST=localhost PORT=4000 pm2 start build/index.js.  

That worked!  I thought I had tried every permutation... but apparently not.  Your comment really made the difference.  Now I can go to my two different urls, tlds, domains and each is serving it's own unique svelte-kit app.  

It still leaves this issue in need of some attention, if even just to include your sentences in the documention, i.e., 

> This looks like is the same trap I fell into. The svelte.config.js env variables are not intended for setting server environment variable values. They tell svelte which env variable names to use. So in svelte.config.js, ...The above tells svelte to look for the env variable HOST and PORT. These must be strings and on your server in your .env or whatever you would use HOST=localhost PORT=4000

```

**Comment by bluwy** [comment]
We had already made an effort to clarify that in the past. The [docs](https://github.com/sveltejs/kit/tree/master/packages/adapter-node#env) looks pretty clear to me how host and port can be used.

**Comment by sridharraman** [comment]
@xpat I have a similar issue. I am trying to host multiple sveltekit apps in my server. I have an added requirement that the apps run by pm2 are in specific locations. If I use `location /`, I am able to view the app in the browser. But if I use any another location in my nginx config, it doesn't work. Do you face that issue too?

**Comment by Rich-Harris** [comment]
I'm going to close this as `wontfix`, since it's [clearly documented](https://github.com/sveltejs/kit/tree/master/packages/adapter-node#environment-variables) and it's really not that onerous to type this...

```
PORT=4567 node build
```

...instead of this:

```
node build
```

If we start exposing default options for port and host in config, then we kinda have to do it for everything (`PROTOCOL_HEADER`, `XFF_DEPTH` etc). That bloats the surface area and leads people into the 'wait, am I doing this the right way?' state of anxiety that comes from not having a single clearly defined solution — the costs outweigh the benefits.
